WASM Demo

ashu26jha opened the pull request:
This Pull Request is intended to shift our WASM demo from using `create_compiled` to dynamically load locales whenever request from a front end. It achieves falling task:

1. Created a `DataProviderManager` which forks locales on the go
2. Added `fallback` with config
3. It prints the fallback locale, also printing trying possibilties
     
    - Instead of logging in the console, we display onto the frontend if. It could a better user experence. This needs approval.

Fixes #2769 

Also we could also discuss the possiblity of adding `DataGen` in our FFI declarations, I could only find it for Dart, not for C++ and JavaScript. I could work on this, if this is what dev wants. This could be useful reduce application's binary size, for eg. 

1. Retrieve user's system locale.
2. Generate the data dynamically for that locale only.

sffc reviewed (commented):
Thanks for looking into this!

ashu26jha commented on docs/tutorials/npm/src/ts/data-provider-manager.ts:
```diff
@@ -0,0 +1,124 @@
+        let fallbackerWithConfig = fallbacker.for_config(fallbackConfig);
+
+        this.fallbacker = fallbackerWithConfig;
+        enProvider.enable_locale_fallback_with(this.fallbacker);
```
Actually I wanted to display the `fallbacked` locale. `ICU4XLocaleFallbackerWithConfig` gives an iterator which helps to get the fallback, not sure how to achieve this using `ICU4XLocaleFallbacker`.

Also it doesn't through an error 😕

robertbastian replied:
The resolved locale is returned in `DataResponseMetdata.locale` (`None` if it's the request locale).

ashu26jha replied:
Just to test the new fallbacker instead of `ICU4XLocaleFallbackerWithConfig`. I tried to write this function, but couldn't create the formatter:

```javascript

let enProvider = await createDataProviderFromBlob('dist/en.postcard');
let unProvider = await createDataProviderFromBlob('dist/und.postcard');
let fallbacker = ICU4XLocaleFallbacker.create(unProvider);
enProvider.enable_locale_fallback_with(fallbacker);
const locale = result(() => ICU4XLocale.create_from_string("bn"));
const groupingStrategy = ICU4XFixedDecimalGroupingStrategy.Auto;
const formatter = result(() => ICU4XFixedDecimalFormatter.create_with_grouping_strategy(
    enProvider,
    unwrap(locale),
    groupingStrategy
));

robertbastian replied:
(oh this is not Rust, ignore my comment).

ashu26jha commented on docs/tutorials/npm/package.json:
```diff
@@ -5,7 +5,7 @@
   "scripts": {
     "clean": "rm dist/*",
-    "build": "webpack",
+    "build": "sh gen.sh && webpack ",
```
Agreed. Noted

ashu26jha commented on docs/tutorials/npm/src/ts/date-time.ts:
```diff
@@ -37,9 +53,31 @@ export class DateTimeDemo {
+    async setLocale(locid: string): Promise <void> {
+        this.#locale = result(() => ICU4XLocale.create_from_string(locid));
+        if (this.#locale.ok == true) {
+            const locales = this.#dataProviderManager.getLoadedLocales();
+            const localesFinal: string[] = [];
+            locales.forEach((item: ICU4XLocale) => {
+                localesFinal.push(item.to_string)
+            })
+            const loadedLocales = new Set(localesFinal);
+            if (!loadedLocales.has(this.#locale.value.to_string())) {
+                await this.updateProvider(this.#locale.value);
+            }
```
Makes sense would reduce repetitive lines 👍

ashu26jha commented on docs/tutorials/npm/src/ts/data-provider-manager.ts:
```diff
@@ -0,0 +1,124 @@
+        const fallbackLocale = await this.trackLocaleFallback(this.fallbacker, icu4xLocale);
+        const newFilePath = `dist/${fallbackLocale}.postcard`;
+        let newProvider = await this.createDataProviderFromBlob(newFilePath);
+        await newProvider.fork_by_locale(this.dataProvider);
```
I tried it, I recieve this msg:

```
ICU4X data error: This provider has been destroyed
```

sffc replied:
The argument passed into fork_by is destroyed, not the receiver.

https://unicode-org.github.io/icu4x/cppdoc/classICU4XDataProvider.html#a5e6150fc8c8dfbeafb43aa29ab5722d2

ashu26jha commented:
Just to test the new fallbacker instead of `ICU4XLocaleFallbackerWithConfig`. I tried to write this function, but couldn't create the formatter:

```javascript

let enProvider = await createDataProviderFromBlob('dist/en.postcard');
let unProvider = await createDataProviderFromBlob('dist/und.postcard');
let fallbacker = ICU4XLocaleFallbacker.create(unProvider);
enProvider.enable_locale_fallback_with(fallbacker);
const locale = result(() => ICU4XLocale.create_from_string("bn"));
const groupingStrategy = ICU4XFixedDecimalGroupingStrategy.Auto;
const formatter = result(() => ICU4XFixedDecimalFormatter.create_with_grouping_strategy(
    enProvider,
    unwrap(locale),
    groupingStrategy
));

ashu26jha commented on tutorials/npm/src/ts/data-provider-manager.ts:
```diff
@@ -0,0 +1,87 @@
+        const enFilePath = 'dist/en.postcard';
+        let enProvider = await this.createDataProviderFromBlob(enFilePath);
+        this.loadedLocales.add(ICU4XLocale.create_from_string("en"));
+        const unFilePath = 'dist/en.postcard';
```
Tried this with `dist/und.postcard`, with no help. Will fix this in consecutive commits**

ashu26jha commented:
Added changes suggested by @sffc & @robertbastian except the `fallbacker`. I have added a sample code of how I am planning to integrate. Is this the correct way to integrate fallbacker?

sffc commented:
ICU4X WG discussion:

- @robertbastian - For splitting locales, this seems reasonable. However, ideally we could also split functionality, but that might result in more duplication.
- @robertbastian - The data provider manager that is being implemented might better live in the FFI crate. We could upstream it later. Might need callbacks.

Conclusion: yes, merge this PR after comments have been resolved.
